### ISSUE TITLE
Colour Contrast detected on button is wrong

### ISSUE THREAD

**Issue by ianbutt** [dequelabs/axe-core#299]
We have an issue where the colour contrast is being picked up incorrectly, the axe report thinks that the background and foreground colour are both white which is incorrect.  The button is actually blue with white letters on a white background.  I have included the json output from axe along with the HTML and css 

Thoughts ?  

"violations": [
        {
            "help": "Elements must have sufficient color contrast",
            "nodes": [
                {
                    "all": [],
                    "impact": "critical",
                    "html": "<span class=\"oj-button-text\" id=\"ui-id-46\">COMPLETE OPERATION</span>",
                    "none": [],
                    "any": [
                        {
                            "id": "color-contrast",
                            "data": {
                                "contrastRatio": "1.00",
                                "fgColor": "#ffffff",
                                "fontSize": "10.5pt",
                                "bgColor": "#ffffff",
                                "fontWeight": "normal"
                            },
                            "message": "Element has insufficient color contrast of 1.00 (foreground color: #ffffff, background color: #ffffff, font size: 10.5pt, font weight: normal)",
                            "impact": "critical",
                            "relatedNodes": []
                        }
                    ],
                    "target": [
                        "#ui-id-46"

HTML is:

<button class="oj-button-lg button-cancel oj-button oj-component oj-enabled oj-default oj-button-full-chrome oj-button-text-only oj-component-initnode" id="complete-operation-button" data-bind="click: onClickCompleteOperation,
                       ojComponent: { component: 'ojButton', label: completeOperationLabel }" aria-labelledby="ui-id-38">
   <div class="oj-button-label">
      <span class="oj-button-text" id="ui-id-38">COMPLETE OPERATION</span>
    </div>
</button> 

the css is:
  $ampButtonColor: #0572CE;

  .amp-button-cancel {
      @include oj-button-state-flat (
          $altButtonTextColor: $ampButtonColor, 
          $altButtonBgColor: #FFFFFF, 
          $altButtonBorderColor: $ampButtonColor, 
          $altButtonTextColorHover: adjust-color($ampButtonColor, $saturation: 20%), // optional
          $altButtonBorderColorHover: adjust-color($ampButtonColor, $saturation: 20%), // optional
          $altButtonTextColorActive: adjust-color($ampButtonColor, $lightness: -15%), // optional
          $altButtonBorderColorActive: adjust-color($ampButtonColor, $lightness: -15%), // optional
          $altButtonBgColorHover: #FFFFFF, // optional
          $altButtonBgColorActive: #FFFFFF // optional
      );
    font-size: 16px;
    margin: auto;
    display: block;
  }

**Comment by ianbutt** [comment]
the html got stripped out:

> <button class="oj-button-lg amp-button-cancel oj-button oj-component oj-enabled oj-default oj-button-full-chrome oj-button-text-only oj-component-initnode" id="complete-operation-button" data-bind="click: onClickCompleteOperation,
                       ojComponent: { component: 'ojButton', label: completeOperationLabel }" aria-labelledby="ui-id-38"><div class="oj-button-label"><span class="oj-button-text" id="ui-id-38">COMPLETE OPERATION</span></div></button>

>

**Comment by ianbutt** [comment]
`<button class="oj-button-lg amp-button-cancel oj-button oj-component oj-enabled oj-default oj-button-full-chrome oj-button-text-only oj-component-initnode" id="complete-operation-button" data-bind="ojComponent: { component: 'ojButton', label: completeOperationLabel }" aria-labelledby="ui-id-38"><div class="oj-button-label"><span class="oj-button-text" id="ui-id-38">COMPLETE OPERATION</span></div></button>`

**Comment by marcysutton** [comment]
It's sort-of hard to tell, do you have a more complete demo in a Codepen or similar, or at least a more complete list of the relevant CSS classes?

**Comment by kbarcham** [comment]
We also have this issue.
I ran aXe on our home page at the iphone 6 viewport.
It's talking about colour contrast issues in our footer which shows white text on a dark background, but is reporting that it is white text on a white background.
https://www.govt.nz/
There is also a real problem in the h1 that I need to fix. please ignore that.

**Comment by marcysutton** [comment]
Thanks @kbarcham, I can reproduce that at a small viewport. There are two incomplete items at the desktop viewport too, that you might want to check. A future version of the aXe Chrome extension will show these items, but for now they can be inspected with the axe-core JavaScript API.

@ianbutt we still need a demo from you please.

**Comment by kbarcham** [comment]
Thanks Marcy.
I was also looking at another NZ government site (www.ird.govt.nz) and noticed that it's picking up the colour contrast issues for the Twitter link (with a score of 3.27), but it's not picking up the more obvious failures of the green "Get it done online All online services" text and the orange "Work it out All calculators and tools" text. Both of which are on the right hand side of the page,
By the way, the orange text on a white background has a luminosity value of 2.91. The green on white fails for small text with a score of 3.49.
I checked this at full screen.

**Comment by mfairchild365** [comment]
Could this be caused by the issue that was fixed in https://github.com/dequelabs/axe-core/pull/272 ?

Maybe we just need a new release that includes that fix?

**Comment by dylanb** [comment]
@kbarcham as far as I can tell, all those "missed" violations are due to background images/gradients. The API exposes these in the "incomplete" array.

**Comment by ianbutt** [comment]
@marcysutton  Here is a code pen that demonstrates everything https://codepen.io/Sarge88/pen/RVWwmW

Let me know if that helps

Ian

**Comment by marcysutton** [comment]
I tested the Codepen just now and it looks resolved with axe-core 2.2.0! Thanks for the contribution.

**Comment by ianbutt** [comment]
@marcysutton  we are using axe-webdriverjs which seems have a dependency on v2.0.7, any chance that we can get a later axe-webdriverjs that has the axe-core of 2.2.0

**Comment by ianbutt** [comment]
I assume we could do this then to use a later axe-core

If you wish to run a specific version of axe-core, you can pass the source axe-core source file in as a string. Doing so will mean axe-webdriverjs runs this version of axe-core, instead of the one installed as a dependency of axe-webdriverjs.
var axeSource = fs.readFileSync('./axe-1.0.js', 'utf8');
var builder = AxeBuilder(driver, axeSource);

**Comment by dylanb** [comment]
@ianbutt you are correct, until the dependency is updated, pass the source in as a string.
